### PR TITLE
Landing page API calls for auction items

### PR DESCRIPTION
**Pull request by Jenn-y**
- Implemented endpoints for frontend landing page: 
- get api/auctions/new_arrivals
- get api/auctions/last_chance

### CODE REVIEW

File: src/main/java/com/auctionapp/api/controller/AuctionController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+@RestController
+@RequestMapping("/api/auctions")
+@AllArgsConstructor
+public class AuctionController {
+
+	private final AuctionService service;
+
+	@GetMapping("/new_arrivals")
+	public ResponseEntity<Collection<AuctionDto>> getNewArrivals() {
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/controller/AuctionController.java]: Why collection instead of list, for example?

File: src/test/java/com/auctionapp/api/model/CategoryTest.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.auctionapp.api.model;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.auctionapp.api.model.dto.CategoryDto;
+
+import org.junit.jupiter.api.Test;
+
+public class CategoryTest {
```
Comment (hanabezdrob-abh) [inline src/test/java/com/auctionapp/api/model/CategoryTest.java:9]: No need to waste your time writing these. We can trust that getters and setters work. In general, if you feel that something makes no sense to test, you are likely right.

File: src/main/java/com/auctionapp/api/service/ItemService.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import com.auctionapp.api.model.entities.Item;
+
+import org.springframework.stereotype.Service;
+
+@Service
+public class ItemService {
+
+	public static Item fromPayload(ItemDto payload) {
+		Item item = new Item();
+		if (payload.getUuid() != null)
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/service/ItemService.java]: Curly brackets are not optional - please don't avoid them

File: src/main/java/com/auctionapp/api/model/entities/Auction.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	private Timestamp startDate;
+
+	@Column(name = "end_date", nullable = true)
+	private Timestamp endDate;
+
+	@Column(name = "highest_bid", nullable = true)
+	private Float highestBid;
+
+	@Column(name = "adress", nullable = true)
+	private String adress;
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/entities/Auction.java]: ~~ typo, `address`

File: src/main/java/com/auctionapp/api/model/entities/StatusEnum.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.auctionapp.api.model.entities;
+
+public enum StatusEnum {
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/entities/StatusEnum.java]: ~~ just Status

File: src/main/java/com/auctionapp/api/model/entities/Item.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@Entity
+@Table(name = "items")
+@Getter
+@Setter
+@NoArgsConstructor
+public class Item extends EntityWithUUID {
+
+	@Column(name = "item_number", nullable = false)
+	private String itemNumber;
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/entities/Item.java]: What's an item number?

**Review by hanabezdrob-abh — COMMENTED** [review]
General guidelines we've discussed before (lombok, final arguments, naming conventions, db types ...) also apply here. Other than those, looks good!

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  subcategory UUID,
+  CONSTRAINT fk_subcategory
+      FOREIGN KEY(subcategory) 
+	  REFERENCES category(id)
+);
+
+CREATE TABLE IF NOT EXISTS item (
+  id UUID DEFAULT gen_random_uuid () PRIMARY KEY,
+  name text NOT NULL,
+  start_price decimal,
```
Comment (hanabezdrob-abh) [inline src/main/resources/db/migration/V2__create_item_cat_auction.sql]: I feel like the start price should be tied to an auction, not an item.

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS auction (
+  id UUID DEFAULT gen_random_uuid () PRIMARY KEY,
+  start_date timestamp,
+  end_date timestamp,
+  highest_bid decimal,
+  address text,
+  zip_code numeric,
+  phone text,
```
Comment (hanabezdrob-abh) [inline src/main/resources/db/migration/V2__create_item_cat_auction.sql]: What are these in an auction? Also, do you think that some other fields should be not null within this table?

File: src/main/resources/db/migration/V1__create_table_user_account.sql [hunk 9]
```diff
@@ -3,12 +3,12 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto;
...
-  lastName text NOT NULL,
+  first_name text NOT NULL,
+  last_name text NOT NULL,
   email text NOT NULL,
   password text NOT NULL,
-  createdAt timestamp,
-  updatedAt timestamp,
+  created_at timestamp,
+  updated_at timestamp,
   role text,
```
Comment (hanabezdrob-abh) [inline src/main/resources/db/migration/V1__create_table_user_account.sql]: Should these be nullable?

File: src/main/java/com/auctionapp/api/model/dto/CategoryDto.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
+package com.auctionapp.api.model.dto;
+
+import java.util.Set;
+import java.util.UUID;
+
+public class CategoryDto {
+	private UUID id;
+	private String name;
+	private CategoryDto subcategory;
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/dto/CategoryDto.java]: Can't a category have several subcategories, not just one?

File: src/main/java/com/auctionapp/api/model/dto/UserDto.java [hunk 11]
```diff
@@ -56,4 +62,40 @@ public Timestamp getCreatedAt() {
...
+		return role;
+	}
+
+	public void setRole(UserRole role) {
+		this.role = role;
+	}
+
+	public void setId(UUID id) {
+		this.id = id;
+	}
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/dto/UserDto.java]: Getters and setters should be grouped (`getX` then `setX`, etc.). Curious as to why you need setters in a DTO?

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  status text,
+  shipping_cost_included text,
+  item_id UUID,
+  seller_id UUID,
+  category_id UUID,
```
Comment (hanabezdrob-abh) [inline src/main/resources/db/migration/V2__create_item_cat_auction.sql]: The category should be tied to an item, not an auction.

File: src/main/java/com/auctionapp/api/model/entities/Category.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	private String name;
+
+	@ManyToOne
+	@JoinColumn(name = "subcategory", referencedColumnName = "id")
+	private Category subcategory;
+
+	@OneToMany(mappedBy = "category", fetch = FetchType.LAZY)
+	private Set<Auction> itemList;
+
+	public Category(){
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/entities/Category.java]: ~~ missing space

File: src/main/java/com/auctionapp/api/service/UserService.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+	public static UserDto toPayload(User user) {
+		UserDto payload = new UserDto();
+		payload.setId(user.getUuid());
+		payload.setEmail(user.getEmail());
+		payload.setFirstName(user.getFirstName());
+		payload.setLastName(user.getLastName());
+		payload.setPassword(user.getPassword());
+		payload.setCreatedAt(user.getCreatedAt());
+		payload.setUpdatedAt(user.getUpdatedAt());
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/service/UserService.java]: Make an appropriate constructor in `UserDto` instead of invoking setters one by one

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  color text,
+  size numeric,
+  description text NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS auction (
+  id UUID DEFAULT gen_random_uuid () PRIMARY KEY,
+  start_date timestamp,
+  end_date timestamp,
+  highest_bid decimal,
```
Comment (hanabezdrob-abh) [inline src/main/resources/db/migration/V2__create_item_cat_auction.sql]: The auction should have a list of all bids tied to it, not just the value of the highest one. It is a logical requirement (and one we will have in the future I believe) to show a list of all bids to the seller.

Comment (Jenn-y) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: There is a separate table named "bid" that tracks the records of all bids.
I put this field here to not have to do constant querying of the bid table each time I have to show the highest bid, but rather this field gets updated when the new higher bid is entered and is always ready for display.

Is it okay to have it like that?

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql (hunk 12 above)
Comment (Jenn-y) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: I was thinking about that. Initially, I tied it to the item table. But when I was working on the landing page, I realized records of items will never be displayed on the platform as such, but rather records of auctions that represent those items will be the ones for display. So then I tied auctions with category to not have to do constant querying over the table of items to fetch the required category as we actually need to categorize auctions on the platform.

What do you think, should I still change it back to the item table?

File: src/main/java/com/auctionapp/api/model/dto/CategoryDto.java (hunk 10 above)
Comment (Jenn-y) [reply src/main/java/com/auctionapp/api/model/dto/CategoryDto.java]: Maybe I should rename that field as subcategory_of, this way it is misleading.
Because actually each subcategory will be created as a record of the category table and will have this field filled with the parent category id.

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql (hunk 8 above)
Comment (Jenn-y) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: Now I remembered that we discussed this when I created the initial ERD, whether the user will have to enter these details for every shipment, if yes, I should put it in the payment details table, if it needs to enter it just once, I should put it in the user account table. You mentioned you would check this, as it is unclear in the design files.

In which table should I move these fields then?

Comment (hanabezdrob-abh) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: I'll get back to you on that as soon as I have all the information necessary. For now, they can be removed completely, because you can add columns to existing tables through subsequent migrations if necessary and I don't believe you need to display these anywhere as of right now.

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql (hunk 15 above)
Comment (hanabezdrob-abh) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: No, because of a few reasons:

- this field has no constraints set on it. If its existence was necessary, it would need to be coupled with the bid table to ensure its value is valid. As it stands right now, someone can maliciously enter a random value into it and the system has no way to react to that as it's just a random standalone field.
- Unless we are talking about significant performance impacts and performance-critical applications, clarity and logic trump making these shortcuts every time. 
- If your auction has a list of all its bids, you could use an `order by` in the API call to order them by price which would put the highest bid as the first every time without polluting the database. You could also have a separate getter in the auction table which could do the same thing.

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql (hunk 8 above)
Comment (Jenn-y) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: Okaay

File: src/main/resources/db/migration/V2__create_item_cat_auction.sql (hunk 15 above)
Comment (Jenn-y) [reply src/main/resources/db/migration/V2__create_item_cat_auction.sql]: Okay, I understand. I deleted the column in this PR, and I will implement it in the new way with the bidding table in the bidding_process branch.

File: src/main/java/com/auctionapp/api/model/dto/UserDto.java [hunk 16]
```diff
@@ -3,14 +3,20 @@
...
-	private final Timestamp createdAt;
-	private final Timestamp updatedAt;
+	private UUID id;
+	private String firstName;
+	private String lastName;
+	private String email;
+	private String password;
+	private Timestamp createdAt;
+	private Timestamp updatedAt;
+	private UserRole role;
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/dto/UserDto.java]: Why did you remove `final` here?

File: src/main/java/com/auctionapp/api/model/entities/User.java [hunk 17]
```diff
@@ -18,119 +18,92 @@
...
-		return createdAt;
-	}
+    public String getFirstName() {
+        return firstName;
+    }
 
-	public void setCreatedAt(Timestamp createdAt) {
-		this.createdAt = createdAt;
-	}
+    public String getLastName() {
```
Comment (hanabezdrob-abh) [inline src/main/java/com/auctionapp/api/model/entities/User.java:86]: Why did you remove setters from the entity? If the user were to edit their profile information once the profile page is implemented, how would you go about that?

File: src/main/java/com/auctionapp/api/model/dto/UserDto.java (hunk 16 above)
Comment (Jenn-y) [reply src/main/java/com/auctionapp/api/model/dto/UserDto.java]: I honestly don't remember purposefully deleting it, it might be that I did it while trying to fix the bug I had with registration and login, so I was checking everything that was different from the last working commit. I put back final now.

File: src/main/java/com/auctionapp/api/model/entities/User.java (hunk 17 above)
Comment (Jenn-y) [reply src/main/java/com/auctionapp/api/model/entities/User.java:86]: Is there a reason why wouldn't I be able or maybe shouldn't add them later, once they are needed during the work on the edit user profile info ticket?

Comment (hanabezdrob-abh) [reply src/main/java/com/auctionapp/api/model/entities/User.java:86]: The reason why you shouldn't is that entities should have setters, however you definitely can add it in later too from a functional standpoint. They are just very easy to add if using a tool such as IntelliJ because they can be generated automatically so there isn't really a need to wait. Either way is fine by me - and by that I mean you can merge this if you won't be fixing it right now, but keep it in mind.
